---
title: "IBD_Analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{IBD_Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
knitr::opts_chunk$set(include = FALSE)
devtools::load_all()
devtools::document()
```

## Data Loading

For this analysis, we will use the `ibd.count` and `ibd.clin` datasets. The `ibd.count` dataset contains the count data for the genes, while the `ibd.clin` dataset contains the clinical data for the patients. Let's load the data first.

```{r loadData}
# Load the required libraries
library(skimr)
library(ggfortify)
library(future)
# Load data
data(ibd.count, ibd.clin)
```

The dataset contains data from 5 different studies concerning inflammatory bowel disease (IBD) in pediatric and adult cohorts publicly available at:

1.  PRJNA248469: <https://www.ncbi.nlm.nih.gov/bioproject/248469>

2.  PRJNA565216: <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE137344>

3.  PRJNA702434: <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE166925>

4.  PRJNA728646: <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE174159>

5.  PRJNA985602: <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE235236>

### Study 1: PRJNA248469

```{r}
# Select first study and remove all NA columns
study1 <- ibd.clin[ibd.clin$Study == "PRJNA248469", ]
# Remove columns that contain only NA values
study1 <- study1[, colSums(is.na(study1)) != nrow(study1)]
# Display the first few rows
head(study1)
```

```{r include=FALSE}
# Transform into factor the columns from 2 to 10 all at once
study1[, c(2,3,5:10)] <- lapply(study1[, c(2,3,5:10)], as.factor)
```

This study contains data from a pediatric cohort of 322 patients of which 218 affected by Crohn Disease (CD), 62 by Ulcerative Colitis (UC) and 42 Normal control (Not IBD)

```{r}
# Use skimr to get a summary of the data
skimr::skim(study1)
```

The tissue from which the samples were extracted is ileal, and RNA-seq assays were performed for all samples. The sex of the patients was recorded, with 187 males and 135 females. For patients with Crohnâ€™s Disease (CD), the ulcer status (deep_ulcera) was also documented, with 76 patients presenting deep ulcers and 142 without deep ulcers. Lastly, we note that the average age of the patients is approximately 12 years, ranging from 2 to 17 years.

We now proceed to the analysis of gene expression data. A Principal Component Analysis (PCA) is performed to evaluate whether the data can be separated based on the clinical condition (class). First, genes with zero variance are filtered out, and then PCA is conducted on the log2-transformed data.

```{r study1PCA, fig.align='center', fig.width=7, fig.height=5}
# select from ibd.count only samples from study1
study1.count <- ibd.count[rownames(ibd.count) %in% rownames(study1), ]
# Remove genes (columns) with zero variance
study1.count <- study1.count[, apply(study1.count, 2, var) != 0]
# PCA of log2-transformed data
pca <- prcomp(log2(study1.count + 1), scale = TRUE)
# Plot the PCA
autoplot(pca, data = cbind(study1.count, study1), 
         colour = "class", shape = "gender") +
  theme_minimal()
```

#### CD vs Not IBD

Proviamo a lanciare la pipeline per il confronto tra CD e Not IBD. Per prima cosa aggiustiamo i dati.

```{r Study1_CDvsNIBD}
# Select only CD and Not IBD patients
s1_CDvsNIBD <- study1[study1$class %in% c("CD", "Not IBD"), ]
# Remove dropped levels from class factor
s1_CDvsNIBD$class <- droplevels(s1_CDvsNIBD$class)
# and do the same for count table
s1.count_CDvsNIBD <- study1.count[rownames(study1.count) %in% rownames(s1_CDvsNIBD), ]

# # Create a new batch column given by the interaction among gender, deep_ulcer and other
# s1_CDvsNIBD$batch2 <- 
#   apply(s1_CDvsNIBD[, c("gender", "deep_ulcer", "other")], 1, function(x) {
#     paste(na.omit(x), collapse = "_")})
# # Transform batch2 into a factor and set levels names
# s1_CDvsNIBD$batch2 <- as.factor(s1_CDvsNIBD$batch2)
# levels(s1_CDvsNIBD$batch2) <- c("F", "F_DU_MaI", "F_NDU_MaI", "F_NDU_MiI",
#                                 "F_NDU_N", "F_NDU_U", 
#                                 "M", "M_DU_MaI", "M_NDU_MaI", "M_NDU_MiI",
#                                 "M_NDU_N", "M_NDU_U")
```

A questo punto possiamo procedere con la pipeline per il confronto tra CD e Not IBD. Per prima cosa, eseguiamo il preprocessing dei dati utilizzando la funzione `preProcess` di `genei`.

```{r Study1_CDvsNIBD_preProcess, fig.align='center', fig.width=7, fig.height=5}
# Preprocess data with preProcess function
pre_proc <- preProcess(
  df.count = s1.count_CDvsNIBD, df.clin = s1_CDvsNIBD, 
  class = "class", batch = "gender", data.type = "rnaseq")
```

The preProcess function returns an object of class preProcess that contains the preprocessed data and intermediate results. Specifically, it includes:

-   The raw count table of the input data.

-   Count tables after each preprocessing step, such as normalization and batch correction.

-   The clinical metadata table provided as input.

This object serves as a structured container, enabling seamless integration with downstream analysis. For instance, you can directly use the preProcess object as input to the runClassifiers function to perform comparative classification analysis between Crohn's Disease (CD) and Not IBD (non-inflammatory bowel disease).

```{r Study1_CDvsNIBD_runClassifiers, fig.align='center', fig.width=7, fig.height=5}
# Increase max size for parallel processing
options(future.globals.maxSize = 2 * 1024^3)
# Define models to run and recipes to use
models <- c(
  'xgboost', 'bag_tree', 'lightGBM', 'pls', 'logistic', 'C5_rules', 
  'mars', 'bag_mars', 'mlp', 'bag_mlp', 'decision_tree', 'rand_forest')#, 
  #'svm_linear', 'svm_poly', 'svm_rbf')
selector.recipes <- c("corr")
# Run classifiers
enso <- runClassifiers(
  preProcess.obj = pre_proc, models = models, selector.recipes = selector.recipes, 
  tuning.method = "tune_grid", n = 3, v = 3, metric = "accuracy", nsim = 2, seed = 123)

```
